spec.html

```diff
@@ -12,194 +12,174 @@ <h1>Locale Objects</h1>
+        1. If _keywords_ does not contain an element whose [[Key]] is the same as _key_, then
+          1. Append the Record{[[Key]]: _key_, [[Value]]: _value_} to _keywords_.
+        1. Let _result_ be a new Record.
+        1. Let _newExtension_ be `"-u"`.
+        1. Let _entry_ be the element of _keywords_ whose [[Key]] is the same as `"attributes"`.
+        1. If _entry_.[[Value]] is not the empty String, then
+          1. Let _newExtension_ be the string-concatenation of _newExtension_, `"-"`, and _entry_.[[Value]].
+        1. Remove _entry_ from _keywords_.
+        1. Repeat for each element _key_ of _relevantExtensionKeys_ in List order,
+          1. Let _value_ be *undefined*.
+          1. If _keywords_ contains an element whose [[Key]] is the same as _key_, then
+            1. Let _entry_ be the element of _keywords_ whose [[Key]] is the same as _key_.
+            1. Let _value_ be _entry_.[[Value]].
+            1. Remove _entry_ from _keywords_.
+          1. Assert: _options_ has a field [[<_key_>]].
+          1. Let _optionsValue_ be _options_.[[<_key_>]].
+          1. If _optionsValue_ is not *undefined*, then
+            1. Let _value_ be _optionsValue_.
+          1. Set _result_.[[<_key_>]] to _value_.
+          1. If _value_  is not *undefined*, then
+            1. Assert: Type(_value_) is String.
+            1. Let _newExtension_ be the string-concatenation of _newExtension_, `"-"`, and _key_.
+            1. If _value_ is not the empty String, then
+              1. Let _newExtension_ be the string-concatenation of _newExtension_, `"-"`, and _value_.
+        1. Repeat for each element _entry_ of _keywords_ in List order,
+          1. Let _key_ be _entry_.[[Key]].
+          1. Let _value_ be _entry_.[[Value]].
+          1. Let _newExtension_ be the string-concatenation of _newExtension_, `"-"`, and _key_.
+          1. If _value_ is not the empty String, then
+            1. Let _newExtension_ be the string-concatenation of _newExtension_, `"-"`, and _value_.
+        1. Let _locale_ be the String value that is _tag_ with all Unicode locale extension sequences removed.
+        1. If the number of elements in _newExtension_ is greater than 2, then
+          1. Let _privateIndex_ be ! Call(%StringProto_indexOf%, _locale_, &laquo; `"-x-"` &raquo;).
+          1. If _privateIndex_ = -1, then
+            1. Let _locale_ be the concatenation of _locale_ and _newExtension_.
+          1. Else,
+            1. Let _preExtension_ be the substring of _locale_ from position 0, inclusive, to position _privateIndex_, exclusive.
+            1. Let _postExtension_ be the substring of _locale_ from position _privateIndex_ to the end of the string.
+            1. Let _locale_ be the string-concatenation of _preExtension_, _newExtension_, and _postExtension_.
+          1. Assert: ! IsStructurallyValidLanguageTag(_locale_) is *true*.
+          1. Let _locale_ be ! CanonicalizeLanguageTag(_locale_).
```

Comment:
Are the previous two lines intended to be indented, or should they also run for locales without a Unicode extension key?

Reply:
No, this part is directly copied from [ResolveLocale, step 2](https://tc39.github.io/ecma402/#sec-resolvelocale). For the final inclusion into the spec we probably want to create a separate abstract operation which then can be shared by `ApplyUnicodeExtensionToTag` and `ResolveLocale`.

Reply:
Do you mean Step 9?

OK, I see that it's the same as what's there. I don't quite understand what it does--do we canonicalize things without an extension?

Reply:
The canonicalization step is just there to ensure the Unicode extension sequence is at the correct canonical position. 

If we have "en-y-not-registrd-ext-but-possible" and then simply append, let's say "-u-ca-gregory", we'd get "en-y-not-registrd-ext-but-possible-u-ca-gregory". But that's not a canonical language tag, because the "y" singleton appears before the "u" singleton. So we need to canonicalize it again and then we get the expected result "en-u-ca-gregory-y-not-registrd-ext-but-possible".

And yes, this case is a bit ridiculous. 😄